### ISSUE TITLE
Fatal signal 11 (SIGSEGV), code 2 (SEGV_ACCERR)

### ISSUE THREAD

**Issue by eliahw** [Gamua/Adobe-Runtime-Support#525]
## Problem Description

Android Crash, primarily on Samsung Galaxy Tab devices with Android 10.
Occured in older Air SDK versions (33.0.x) up to current version (33.1.1.300).

We have several customers with this problem, updating Air SDK did not help.

**Air SDK:** 33.1.1.300

```
Fatal signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0xc34f6000 in tid 2030 (lag.abc1android), pid 2030 (lag.abc1android)
pid: 2030, tid: 2030, name: lag.abc1android  >>> air.de.mildenbergerverlag.abc1android <<<
       #00 pc 00ceb000  /data/app/air.de.mildenbergerverlag.abc1android-OUY41DIGeYjbWthk3Kx9Vw==/lib/arm/libCore.so!libCore.so (offset 0xce7000) (BuildId: ca591d11ca1f5800fa560b863774bca9a69742cc)
```

## Steps to Reproduce

Any build while trying to fix the error led to the same error (with offset 0xce7000 in libCore.so) on our Samsung Galaxy Tab 10.1 WiFi.
We also tried removing all ANEs, all extensions/librarys and all code in the main entry point of the ActionScript file with no success.

Other devices, including an Android 11 phone are not affected by this crash.

**Comment by Gokulv617** [comment]
AIR 33.1.1.259
It is happening in all my apps also. Please check the attached screenshots for more details

![501-p1-signal 11 crash](https://user-images.githubusercontent.com/43432782/98904883-71799600-24e0-11eb-90dc-93554f6e90cb.png)


![501-p1-signal 11 users](https://user-images.githubusercontent.com/43432782/98904762-3e370700-24e0-11eb-9b51-d43864b46ece.png)

**Comment by Kelsonk** [comment]
I'm also working through a huge increase in crashes with this code. Can anyone tell me what the latest air version that works is? I'll post screens and such if requested.

**Comment by Gokulv617** [comment]
**UPDATED from AIR 33.1.1.259 to AIR 33.1.1.300**
**Another type of the signal 11 crash with <anonymous> tag happens a lot more in 64 bit apk. This crash has increased a lot in this new AIR version. For the last 7 days I have faced around 700 crashes in this single app and it is increasing every hour. Please check the logs below** Please suggest something to reduce this crash

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 0 >>> air.com.app.application<<<

backtrace:
  #00  pc 0000000000034038  <anonymous>
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***

![501-p2 - AIR 33 1 1 300 - 64 bit](https://user-images.githubusercontent.com/43432782/99213088-b6f4d680-27f2-11eb-96dd-15724a238c88.png)

![graph-AIR 331 1 300-64bit](https://user-images.githubusercontent.com/43432782/99213452-b27ced80-27f3-11eb-8b7e-5aa90a319233.png)


![devices - 501 p2 - 33 1 1 300 - 64 bit](https://user-images.githubusercontent.com/43432782/99213102-bd834e00-27f2-11eb-8381-90443b54288a.png)

**Comment by Kelsonk** [comment]
Jest here to repeat what Gokulv617 said. Although it's a bit early, it does look like going back to 259 removes these crashing issues.

**Comment by falcos** [comment]
I got the same problem with AIR 33.1.1.300. 
Kelsonk, did anything improve with going back to AIR 33.1.1.259?

**Comment by Gokulv617** [comment]
yes the crashes happened in my app is reduced going back AIR 33.1.1.259

![501-p2](https://user-images.githubusercontent.com/43432782/99524188-118e5e00-29be-11eb-9390-9964554f870c.png)


> I got the same problem with AIR 33.1.1.300.
> Kelsonk, did anything improve with going back to AIR 33.1.1.259?

**Comment by Kelsonk** [comment]
> I got the same problem with AIR 33.1.1.300.
> Kelsonk, did anything improve with going back to AIR 33.1.1.259?

Not sure if crashes are going to go down to the same rate as they were when I was using 33.0.x, but yes, you can see when I tried 300 and went back to 259.

![crashes](https://user-images.githubusercontent.com/6390370/99559802-0580ad00-297a-11eb-885b-76e8c9a046c5.png)

Are there any updates here? I'm not sure if this affects iOS as well but I can't build with an earlier version of the sdk on iOS due to a lack of iOS 14.

**Comment by falcos** [comment]
I also updated the apps with AIR 33.1.1.259 and the crash with this code is gone for now. I had yesterday 20k+ reports of it.

**Comment by Kelsonk** [comment]
Does anyone have any info on iOS? I'd like to do an update but don't know whether it matters if I use the newest SDK

**Comment by ajwfrost** [comment]
Hi

So we've been looking into this, with the above information.. and a big difference in crashes for a particular version is obviously a red flag. The crash locations are odd though and not in our code, which meant it was a likely JIT issue (i.e. crash in generated code). There was a jit change that went into the .300 build so we have been re-analysing that and believe we've found an issue that could cause an invalid program counter i.e. PC moves to some memory that perhaps we don't own (or isn't set as executable) hence causing a seg fault.

We need to do a little more analysis and try to confirm a fix here, should then be something we put into our next release before too long..

This means that the issue is only happening during run-time JIT, so for iOS platforms (AOT-based) this wouldn't have an impact.

thanks

  Andrew

**Comment by Kelsonk** [comment]
> Hi
> 
> So we've been looking into this, with the above information.. and a big difference in crashes for a particular version is obviously a red flag. The crash locations are odd though and not in our code, which meant it was a likely JIT issue (i.e. crash in generated code). There was a jit change that went into the .300 build so we have been re-analysing that and believe we've found an issue that could cause an invalid program counter i.e. PC moves to some memory that perhaps we don't own (or isn't set as executable) hence causing a seg fault.
> 
> We need to do a little more analysis and try to confirm a fix here, should then be something we put into our next release before too long..
> 
> This means that the issue is only happening during run-time JIT, so for iOS platforms (AOT-based) this wouldn't have an impact.
> 
> thanks
> 
> Andrew

Thanks for the update!

**Comment by eliahw** [comment]
> Hi
> 
> So we've been looking into this, with the above information.. and a big difference in crashes for a particular version is obviously a red flag. The crash locations are odd though and not in our code, which meant it was a likely JIT issue (i.e. crash in generated code). There was a jit change that went into the .300 build so we have been re-analysing that and believe we've found an issue that could cause an invalid program counter i.e. PC moves to some memory that perhaps we don't own (or isn't set as executable) hence causing a seg fault.
> 
> We need to do a little more analysis and try to confirm a fix here, should then be something we put into our next release before too long..
> 
> This means that the issue is only happening during run-time JIT, so for iOS platforms (AOT-based) this wouldn't have an impact.
> 
> thanks
> 
> Andrew

Hey Andrew, thanks for the update!

As mentioned in my first post, for us the issue first occured with an app built with 33.0.2.338 almost exclusively on Samsung Galaxy Tab devices with Android 10.

While trying to find the problem, we rebuilt our app with 33.1.1.300 and the problem occured on other devices too, which you can also see by other replies in this issue.
As we hope you can trace down the issue with .300, we are still not sure if this is the same issue happening on Samsung Galaxy Tab.

**Comment by ajwfrost** [comment]
@eliahw  yes I had wondered about this, we may be looking at two different issues here. But to clarify that original post, can you please confirm what version of AIR that segfault information was taken from? 
```      #00 pc 00ceb000  /data/app/air.de.mildenbergerverlag.abc1android-OUY41DIGeYjbWthk3Kx9Vw==/lib/arm/libCore.so!libCore.so (offset 0xce7000) (BuildId: ca591d11ca1f5800fa560b863774bca9a69742cc)```

The one thing I wondered is if this was caused by the selinux thing but that was fixed in 33.1.1.217. If you can confirm the build number we can check the symbols and see if there's anything obvious.

I suspect moving to 33.1.1.300 exposed the other issues too.. a bit confusing when a 'segv_accerr' can come from many different causes and locations!

thanks

**Comment by Mintonist** [comment]
For information:
![image](https://user-images.githubusercontent.com/9267850/99715372-d1130b00-2ab7-11eb-8073-a7fe7b4a8040.png)

**Comment by eliahw** [comment]
@ajwfrost The segfault in my original post was from the new build with Air SDK 33.1.1.300. I also found some more information in the logs that happened shortly before crashing.
Just for comparison, here are the segfault information for both builds on the same device:

**Air SDK 33.1.1.300:**
```
avc:  denied  { execmod } for  pid=8098 comm="lag.abc1android" path="/data/app/air.de.mildenbergerverlag.abc1android-OUY41DIGeYjbWthk3Kx9Vw==/lib/arm/libCore.so" dev="mmcblk0p33" ino=1376759 scontext=u:r:untrusted_app:s0:c218,c256,c512,c768 tcontext=u:object_r:apk_data_file:s0 tclass=file permissive=0 SEPF_SM-T510_10_0020 audit_filtered
arch=40000028 syscall=125 per=8 success=no exit=-13 a0=b986e000 a1=1000 a2=5 a3=0 items=0 ppid=3546 pid=8098 auid=4294967295 uid=10218 gid=10218 euid=10218 suid=10218 fsuid=10218 egid=10218 sgid=10218 fsgid=10218 tty=(none) ses=4294967295 comm="lag.abc1android" exe="/system/bin/app_process32" subj=u:r:untrusted_app:s0:c218,c256,c512,c768 key=(null)

...

Fatal signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0xb986e000 in tid 8098 (lag.abc1android), pid 8098 (lag.abc1android)

...

pid: 8098, tid: 8098, name: lag.abc1android  >>> air.de.mildenbergerverlag.abc1android <<<
    #00 pc 00ceb000  /data/app/air.de.mildenbergerverlag.abc1android-OUY41DIGeYjbWthk3Kx9Vw==/lib/arm/libCore.so!libCore.so (offset 0xce7000) (BuildId: ca591d11ca1f5800fa560b863774bca9a69742cc)
```

**Air SDK 33.0.2.338 (original):**
```
avc:  denied  { execmod } for  pid=19678 comm="lag.abc1android" path="/data/app/air.de.mildenbergerverlag.abc1android-UHxOj6Cu8WbepGr9g_cEPw==/lib/arm/libCore.so" dev="mmcblk0p33" ino=1376626 scontext=u:r:untrusted_app:s0:c219,c256,c512,c768 tcontext=u:object_r:apk_data_file:s0 tclass=file permissive=0 SEPF_SM-T510_10_0020 audit_filtered

...

Fatal signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0xb982e000 in tid 19678 (lag.abc1android), pid 19678 (lag.abc1android)

...

pid: 19678, tid: 19678, name: lag.abc1android>>> air.de.mildenbergerverlag.abc1android <<<
    #00 pc 00ceb000  /data/app/air.de.mildenbergerverlag.abc1android-UHxOj6Cu8WbepGr9g_cEPw==/lib/arm/libCore.so!libCore.so (offset 0xce7000) (BuildId: ca591d11ca1f5800fa560b863774bca9a69742cc)
```

**Comment by ajwfrost** [comment]
@eliahw  thanks that's very useful detail - so it looks like we're getting blocked by the SELinux security when trying to .. modify the executable?! That was something that was happening in earlier build (optimising itself based on CPU characteristics) but was removed in 33.1.1.217 (part of the fixes for #372). There's no text relocation left in the binary, I just double-checked that, so it's a bit of a mystery what is causing this error. 

See 'execmod' reference in https://www.nsa.gov/Portals/70/images/resources/everyone/digital-media-center/publications/research-papers/implementing-selinux-as-linux-security-module-report.pdf

Actually .. having delved a little more into the binary, there is a clue:
00ce7000 l     O .data.rel.ro   00000014              .hidden vtable for SkRBufferWithSizeCheck
So, something in the Skia library perhaps, or the way we're linking that in? We will look into this.... thanks for the extra details!

**Comment by falcos** [comment]
Just an update: our crash rate went from 8% to 0.15% after downgrading from 33.1.1.300 to 33.1.1.259
![aa1](https://user-images.githubusercontent.com/6403159/100514236-9f670780-317b-11eb-919d-09464c4dc7ef.png)

**Comment by alphatwin** [comment]
Hello , thanks for reports. Does anyone know where to get older versions like AIR 33.1.1.259?

**Comment by Gokulv617** [comment]
> Hello , thanks for reports. Does anyone know where to get older versions like AIR 33.1.1.259?

You can get any older version by just editing the AIR version number in the below link

https://airsdk.harman.com/download/33.1.1.259

**Comment by ajwfrost** [comment]
Not sure about “any” version, there were a lot that probably aren’t worth getting ... if you look at the “Release Notes” web page, there are links to the versions that are still available...
http://airsdk.harman.com/release_notes

**Comment by dmitrydmsk** [comment]
Hello, i have the same problem [ signal 11 (SIGSEGV), code 2 (SEGV_ACCERR) libCore.so!libCore.so (offset 0x11a1000) ]
On app update i was forced to set target sdk to 29 (from 28) and crash rate increased by several times, but mostly for 32 bit version (although there is a 1 case so far for 64 bit vesion). 64 bit was compiled with AIR 33.1.1259 and 32 bit with old AIR 32 (i didn't upgrade 32 bit version to higher AIR sdk since it just work.. until now). 
  As i understand, use of 33.1.1259 sdk help to reduce crash rate. Is it true for both versions 64 and 32 bit?

**Comment by eliahw** [comment]
Hello, any news about this problem or something we can do? Our customers are still not able to use our app.

**Comment by teotigraphix** [comment]
@ajwfrost With 300, I have a custom Windows ane that wraps a C++ audio engine and I use a .dll on Windows and fmodex.dll for the driver hookup.

The very first time I compiled my app with 300, it crashed at loading the ane on Windows. Then it said my ane has is not current with either swf version, then the actual AIR manifest version.

I recompiled the ane with 300, the compatible errors went away and then ran the app but crashes with ane Not Support method error.

Did you people change any native library loading type things in 300? I can't develop on Windows and this was the version for Android 29 I needed.

I have used this ane for years without any crashes on my dev Windows 10 setup.

**Comment by teotigraphix** [comment]
@ajwfrost I must have tagged this onto a non relevant thread BUT some good news, the update to 345 solved the library loading problem with my Windows C++ ane. THANK YOU!

**Comment by falcos** [comment]
Does anyone know if 33.1.1.345 comes with a fix for the crash referred to in this thread?

**Comment by ajwfrost** [comment]
If by "the crash" you mean the problem that caused a large spike in crashes in the .300 build, then yes. Build 345 is hopefully preferable to going back to build 259.. I think there are still a couple of open items where we still have some instabilities but the main issues introduced in .300 (JIT and audio) are fixed now.

**Comment by Gokulv617** [comment]
@ajwfrost 
Does 33.1.1.345 comes with a fix for any ANR's, specifically related to Input dispatch and broadcast of intent ?

**Comment by Mintonist** [comment]
@ajwfrost Please close issues if you are resolving them. It's signal to us to check it.

**Comment by denweng** [comment]
Unfortunately the initial problem @eliahw reported with theis issue is not solved.

The problem suddenly appeared in Galaxy Tabs with our up to then good working app. We don't know if any os updates came in this time since we bought the tablet for tests after user reported problems.

With the newer sdks the problem is the same.

**Comment by idanasher** [comment]
Hello everybody

my android app has also started getting LOTS of this crash (Fatal signal 11).
It is happening on Android 10 Device, mainly on Samsung galaxy S-20\+\ultra ...

My app is a bit old, publish with AIR 33.0.2.338

Should I compile it with newest  SDK available 33.1.1.345 ?

Will this fix the crashes , or should I wait for this to be resolved ?

(I did not understand what is the status of this issue)

Thanks

**Comment by ajwfrost** [comment]
@idanasher if you're starting to see issues with an app and you haven't updated the app, then it's more likely to be the OS compatibility. Looking at the versions and details there I would hazard a guess you're being hit by one of the restrictions in Android 10 regarding self-modifying executables, which means yes you need to update to the latest SDK.

Always a challenge to determine which segfaults are caused by which issues, sadly there are multiple ones..

@Mintonist just saw your suggestion there to close issues when we've done a fix; I had thought the other way round, i.e. that the originator should be the only one to close an issue, once they've confirmed it is no longer a problem for them.. but yes we'd be very happy to close down issues when we think they're fixed, it would clean things up a little :-)

**Comment by Mintonist** [comment]
@ajwfrost  " but yes we'd be very happy to close down issues when we think they're fixed, it would clean things up a little :-)"

All community wait for it!) 
Not only author is interested in issues. So reopen in new issue (if needed) is not so difficult)

**Comment by SavedByZero** [comment]
Please don't lock this yet, as this is happening to me also. The folks at [proprietary haxe engine name redacted] found the above SIGSEGV error in their logs.  
 I upgraded from 33.1.1.190 to 33.1.1.406, and this did not seem to fix it; either the app will crash, or it will hang indefinitely with a black screen.  Same thing happens with the other sample projects from their engine and they basically pointed me to this thread.

**Comment by lwmirkk** [comment]
My app had a lot of random ANR's (crash, stuck, black screen...) when I played videos on my app.
After using a video playback ANE, my app never had an ANR again.

**Comment by ajwfrost** [comment]
@SavedByZero are you using multimedia as well?

I suspect for a lot of the simple/standard cases where an ANE can provide the functionality, then AIR is just over-complicating things with a huge number of internal classes and threads going on when all it perhaps needs to do is to use the appropriate native playback engine! Definitely an area we want to focus on immediately after we've moved over to App Bundle and sorted out this year's Google updates...

**Comment by SavedByZero** [comment]
@ajwfrost I don't have any ANEs or videos in my project...I feel like we've been at this point before long ago, though at the time The Engine's updates seemed to fix things.  But they seem inclined to think it's an AIR issue.  My guess is that their library makes use of something in AIR that doesn't work well that other apps might not always use, but that's as far as I can speculate...

**Comment by ajwfrost** [comment]
Ah .. that's ringing vague bells. If there are any call stacks that you have from the sigsegv errors then we can look at them (if we know the full AIR SDK version) and if it's an ANR issue then I would be interested if you could get a Scout log of the application running for a while (even if you don't reproduce the issue .. it would be good to see what's happening with the mm/gc).

And feel free to pass my contact details onto the engine developers in case we could work with them to understand what's going on..

thanks

**Comment by SavedByZero** [comment]
thanks @ajwfrost -- I just connected you and them in an e-mail thread.

**Comment by ventr1x** [comment]
These issues, that have no tracable stack for us at all, get out of hand. Every SDK version I see more and more and more crashes without us changing anything (besides upgrading ANEs here and there).

Our last update with newest AIR (33.1.1.476) doubled(!) our crash rate. That is not fun with a user base of around 400k active installs across 3 apps.

Example that shows sudden new crash entries after update:
![Screenshot 2021-06-01 at 11 37 11](https://user-images.githubusercontent.com/7974355/120302431-2ed9a880-c2ce-11eb-8cd3-3a4ea2696b04.png)

There is no trace besides the typical "libCore.so".

**Comment by ajwfrost** [comment]
@ventr1x is there no information at all about the SEGV_MAPERR happening here? It's a different issue from what's been reported previously under this thread, which as a SEGV_ACCERR, but without any details on a fault address it's going to be difficult to know what's caused it. The only real option would be to build a bunch of intermediary versions to isolate what change caused the difference in frequency of the problem..

I am wondering though at the fact this mirrors the ANR rate quite significantly. I'm not sure whether they're reporting an ANR as a crash, or whether there's a 'crash' that doesn't bring down the process but hangs it until the ANR happens.. (that shouldn't really be possible though).

Can I check what version of the SDK your previous build was using? and are these happening on particular devices, OS versions, CPU architectures, etc?

thanks
